altair/utils/data.py

```diff
@@ -91,34 +114,71 @@ def limit_rows(data, max_rows=5000):
 
 
 @curried.curry
-def sample(data, n=None, frac=None):
+def sample(
+    data: _DataType, n: Optional[int] = None, frac: Optional[float] = None
+) -> Optional[Union[pd.DataFrame, Dict[str, Sequence], "pyarrow.lib.Table"]]:
     """Reduce the size of the data model by sampling without replacement."""
     check_data_type(data)
     if isinstance(data, pd.DataFrame):
         return data.sample(n=n, frac=frac)
     elif isinstance(data, dict):
         if "values" in data:
             values = data["values"]
-            n = n if n else int(frac * len(values))
+            if not n:
+                if frac is None:
+                    raise ValueError(
+                        "frac cannot be None if n is None and data is a dictionary"
+                    )
+                n = int(frac * len(values))
             values = random.sample(values, n)
             return {"values": values}
+        else:
+            # Maybe this should raise an error or return something useful?
+            return None
     elif hasattr(data, "__dataframe__"):
         # experimental interchange dataframe support
         pi = import_pyarrow_interchange()
         pa_table = pi.from_dataframe(data)
-        n = n if n else int(frac * len(pa_table))
+        if not n:
+            if frac is None:
+                raise ValueError(
+                    "frac cannot be None if n is None with this data input type"
+                )
+            n = int(frac * len(pa_table))
         indices = random.sample(range(len(pa_table)), n)
         return pa_table.take(indices)
+    else:
+        # Maybe this should raise an error or return something useful? Currently,
+        # if data is of type SupportsGeoInterface it lands here
+        return None
+
+
+class _JsonFormatDict(TypedDict):
+    type: Literal["json"]
+
+
+class _CsvFormatDict(TypedDict):
+    type: Literal["csv"]
+
+
+class _DataJsonUrlDict(TypedDict):
+    url: str
+    format: _JsonFormatDict
```

Comment:
Should format be optional?

Reply:
So far this is only used as a return type for `to_json` which always returns `format: _JsonFormatDict` so I think it's accurate for that usage. But maybe you mean it could be confused to mean that a data json dictionary always needs format specified? I renamed the classes to make it clearer that these are not general schemas but specific type hints for those functions.

Does that address it?